src/software/thunderscope/robot_diagnostics/diagnostics_widget.py

```diff
@@ -0,0 +1,165 @@
+from pyqtgraph.Qt import QtCore, QtGui
+from pyqtgraph.Qt.QtWidgets import *
+from pyqtgraph.Qt.QtCore import *
+from proto.import_all_protos import *
+from software.logger.logger import create_logger
+
+from software.thunderscope.proto_unix_io import ProtoUnixIO
+from software.thunderscope.robot_diagnostics.chicker_widget import ChickerWidget
+from software.thunderscope.robot_diagnostics.handheld_device_status_view import (
+    HandheldDeviceStatusView,
+    HandheldDeviceConnectionStatus,
+)
+from software.thunderscope.robot_diagnostics.diagnostics_input_widget import (
+    DiagnosticsInputToggleWidget,
+    ControlMode,
+)
+from software.thunderscope.robot_diagnostics.handheld_device_manager import (
+    HandheldDeviceManager,
+)
+from software.thunderscope.robot_diagnostics.drive_and_dribbler_widget import (
+    DriveAndDribblerWidget,
+)
+
+
+class DiagnosticsWidget(QScrollArea):
+    """
+    The DiagnosticsWidget contains all widgets related to diagnostics:
+        - HandheldDeviceStatusWidget
+        - DiagnosticsInputWidget
+        - DriveAndDribblerWidget
+        - ChickerWidget
+
+    This widget is also responsible for controlling the diagnostics inout,
+    either from the DriverAndDribblerWidget or the HandheldDeviceManager,
+    depending on the user selected choice from the DiagnosticsWidget UI
+    """
+
+    # signal to indicate if manual controls should be disabled based on enum mode parameter
+    diagnostics_input_mode_signal = pyqtSignal(ControlMode)
+
+    # signal to indicate that handheld device should be reinitialized
+    reinitialize_handheld_device_signal = pyqtSignal()
+
+    # signal to indicate that handheld device was disconnected
+    handheld_device_connection_status_signal = pyqtSignal(
+        HandheldDeviceConnectionStatus
+    )
+
+    def __init__(self, proto_unix_io: ProtoUnixIO) -> None:
+        super(DiagnosticsWidget, self).__init__()
+
+        self.logger = create_logger("RobotDiagnostics")
+
+        self.proto_unix_io = proto_unix_io
+
+        # default start with diagnostics input
+        self.__control_mode = ControlMode.DIAGNOSTICS
+
+        # default start with disconnected controller
+        self.__handheld_device_status = HandheldDeviceConnectionStatus.DISCONNECTED
+
+        # initialize widgets
+        self.handheld_device_status_widget = HandheldDeviceStatusView(
+            self.reinitialize_handheld_device_signal
+        )
+        self.drive_dribbler_widget = DriveAndDribblerWidget(self.proto_unix_io)
+        self.chicker_widget = ChickerWidget(self.proto_unix_io)
+        self.diagnostics_control_input_widget = DiagnosticsInputToggleWidget(
+            self.diagnostics_input_mode_signal
+        )
+
+        # connect input mode signal with handler
+        self.diagnostics_input_mode_signal.connect(self.__control_mode_update_handler)
```

Comment:
maybe the signal should be connected before passing it into the widget? I could see a possible race condition happen if the signal gets called before line 73 happens